### PR TITLE
lint: enable govet shadow linter and resolve linter warnings

### PR DESCRIPTION
**Pull request by cce**
## Summary

This enables the govet shadow analysis pass, implemented in https://github.com/golang/tools/blob/master/go/analysis/passes/shadow/shadow.go and enabled via golangci-lint.

The shadow linter prevents some cases of shadowed variables being used, which especially impacts the re-use of common variable names like `err` and `ok`. In this PR I attempted to resolve these issues mostly be just renaming names like `err` to more specific and unused names like `err2` or `osErr`.

In a few other cases I updated
```golang
err := X()
if err != nil {
    return err
}
```
to 
```golang
err = X()
if err != nil {
    return err
}
```
if the check-and-return came right away. I also removed a couple of named error return variables or unnecessary declarations of names earlier in a function that were shadowed later.

The shadow linter `strict: false` setting is used which comes with the [following explanation](https://github.com/golang/tools/blob/v0.7.0/go/analysis/passes/shadow/shadow.go#L104-L122):
> A span stores the minimum range of byte positions in the file in which a given variable (types.Object) is mentioned. It is lexically defined: it spans from the beginning of its first mention to the end of its last mention. A variable is considered shadowed (if strict is off) only if the shadowing variable is declared within the span of the shadowed variable. In other words, if a variable is shadowed but not used after the shadowed variable is declared, it is inconsequential and not worth complaining about. This simple check dramatically reduces the nuisance rate for the shadowing
> check, at least until something cleverer comes along.
>
> One wrinkle: A "naked return" is a silent use of a variable that the Span will not capture, but the compilers catch naked returns of shadowed variables so we don't need to.
>
> Cases this gets wrong (TODO):
> - If a for loop's continuation statement mentions a variable redeclared in the block, we should complain about it but don't.
> - A variable declared inside a function literal can falsely be identified as shadowing a variable in the outer function.

## Test Plan

Existing tests should pass — in most cases I did not change the logic at all, but I did make some bigger changes to resolve some of the warnings.

### CODE REVIEW

File: ledger/catchpointtracker.go [hunk 1]
```diff
@@ -1207,21 +1207,20 @@ func (ct *catchpointTracker) recordFirstStageInfo(ctx context.Context, tx tracke
 		return err
 	}
 
-	{
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/catchpointtracker.go:1210]: this is from Tolik's scoping, I removed some but not all

**Comment by gmalouf** [comment]
Talked this through with @cce , basically we are mostly okay with the inconsistent handling in some cases (see cmd/goal/asset.go for an example) as long as the linter (which in ways is stricter than needed) is happy with both of them. 

For example, we are fine re-using an err variable directly, as long as it is not shadowing (and there is a quick nil check) as in https://github.com/algorand/go-algorand/pull/5261/files#diff-43408fd8ee96ef81b41df4152675473a9d2b76358da39757022f1e2d6ee00872R764 whereas in other situation err2, err3, etc is more appropriate for readability.

File: ledger/acctupdates.go [hunk 2]
```diff
@@ -1523,7 +1523,8 @@ func (au *accountUpdates) lookupWithoutRewards(rnd basics.Round, addr basics.Add
 }
 
 // getCreatorForRound returns the asset/app creator for a given asset/app index at a given round
-func (au *accountUpdates) getCreatorForRound(rnd basics.Round, cidx basics.CreatableIndex, ctype basics.CreatableType, synchronized bool) (creator basics.Address, ok bool, err error) {
+func (au *accountUpdates) getCreatorForRound(rnd basics.Round, cidx basics.CreatableIndex, ctype basics.CreatableType, synchronized bool) (
```
[the remark this answers is not in the document]
Comment (gmalouf) [reply ledger/acctupdates.go:1526]: I wish we could enforce a rule around not naming return variables when functions are too long strictly - but doubt linter picks it up.

[the remark this answers is not in the document]
Comment (jannotti) [reply ledger/acctupdates.go:1526]: I don't thinking _naming_ them is the problem, _naked returns_ are the problem.  You might name return values if you are trying to distinguish two returned ints - you want to explain which one is which, in the signature.

But naked returns are confusing. It's so easy to not realize what you're returning, especially in the presence of shadowing.  So I would much rather ban naked returns in long functions.  I very often return things explicitly, even if a naked return would do the same thing. In a long function, it's much easier to understand without a moment's hesitation.

File: ledger/catchpointtracker.go [hunk 3]
```diff
@@ -948,18 +948,17 @@ func (ct *catchpointTracker) close() {
...
 	}
-	var added, deleted bool
 	accumulatedChanges := 0
 
 	for i := 0; i < accountsDeltas.len(); i++ {
 		delta := accountsDeltas.getByIdx(i)
 		if !delta.oldAcct.AccountData.IsEmpty() {
 			deleteHash := trackerdb.AccountHashBuilderV6(delta.address, &delta.oldAcct.AccountData, protocol.Encode(&delta.oldAcct.AccountData))
-			deleted, err = ct.balancesTrie.Delete(deleteHash)
+			deleted, err := ct.balancesTrie.Delete(deleteHash)
```
Comment (gmalouf) [inline ledger/catchpointtracker.go:961]: These are safe without naming the error variable in function signature, frankly more readable too.

File: ledger/catchupaccessor.go [hunk 4]
```diff
@@ -627,9 +627,9 @@ func (c *catchpointCatchupAccessorImpl) processStagingBalances(ctx context.Conte
 			}
 		}
 		if hasCreatables {
-			start := time.Now()
+			writeStart := time.Now()
```
Comment (gmalouf) [inline ledger/catchupaccessor.go:630]: Nit - could call it `writeCreatStart` for consistenct.

File: ledger/catchupaccessor.go [hunk 5]
```diff
@@ -754,7 +754,7 @@ func (c *catchpointCatchupAccessorImpl) BuildMerkleTrie(ctx context.Context, pro
 		defer wg.Done()
 		defer close(writerQueue)
 
-		err := dbs.Snapshot(func(transactionCtx context.Context, tx trackerdb.SnapshotScope) (err error) {
+		dbErr := dbs.Snapshot(func(transactionCtx context.Context, tx trackerdb.SnapshotScope) (err error) {
```
Comment (gmalouf) [inline ledger/catchupaccessor.go:757]: Why didn't we rename/remove the err error declaration in the signature of this function (its pretty long)?

**Review by gmalouf — DISMISSED** [review]

File: ledger/acctupdates.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tzaffi) [reply ledger/acctupdates.go:1512]: It depends on how many failures currently but ideally even less (enough for very simple if / else).

[the remark this answers is not in the document]
Comment (gmalouf) [reply ledger/acctupdates.go:1512]: I was thinking 15-20 lines max - not sure it's a blocker for this PR but a quick follow-up.

File: cmd/loadgenerator/main.go [hunk 6]
```diff
@@ -106,11 +106,11 @@ func main() {
 	if (cfg.ClientURL == nil || cfg.ClientURL.String() == "") || cfg.APIToken == "" {
 		if algodDir != "" {
 			path := filepath.Join(algodDir, "algod.net")
-			net, err := os.ReadFile(path)
-			maybefail(err, "%s: %v\n", path, err)
+			net, osErr := os.ReadFile(path)
+			maybefail(err, "%s: %v\n", path, osErr)
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply cmd/loadgenerator/main.go:110]: Yes, but the second error argument in `maybefail` is only used for logging, not for logic.

File: cmd/goal/account.go [hunk 7]
```diff
@@ -948,16 +948,16 @@ No --delete-input flag specified, exiting without installing key.`)
...
+			vErr = fmt.Errorf("unable to verify key installation. Verify key installation with 'goal account partkeyinfo' and delete '%s', or retry the command. Error: %w", partKeyFile, vErr)
+			reportErrorf(errorRequestFail, vErr)
 		}
 
 		reportInfof("Participation key installed successfully, Participation ID: %s\n", addResponse.PartId)
 
 		// Delete partKeyFile
-		if nil != os.Remove(partKeyFile) {
-			reportErrorf("An error occurred while removing the partkey file, please delete it manually: %s", err)
+		if osErr := os.Remove(partKeyFile); osErr != nil {
```
Comment (tzaffi) [inline cmd/goal/account.go:959]: I prefer this pattern, but is this an opinionated change, or enforced by the new linter config?

File: cmd/loadgenerator/main.go [hunk 8]
```diff
@@ -106,11 +106,11 @@ func main() {
...
 			path = filepath.Join(algodDir, "algod.token")
-			token, err := os.ReadFile(path)
-			maybefail(err, "%s: %v\n", path, err)
+			token, osErr := os.ReadFile(path)
+			maybefail(err, "%s: %v\n", path, osErr)
```
Comment (tzaffi) [inline cmd/loadgenerator/main.go:113]: ^^^ similar issue ^^^

File: data/transactions/logic/eval.go [hunk 9]
```diff
@@ -2625,26 +2625,26 @@ func opRetSub(cx *EvalContext) error {
 	if top < 0 {
 		return errors.New("retsub with empty callstack")
 	}
-	frame := cx.callstack[top]
-	if frame.clear { // A `proto` was issued in the subroutine, so retsub cleans up.
-		expect := frame.height + frame.returns
+	topFrame := cx.callstack[top]
```
Comment (tzaffi) [inline data/transactions/logic/eval.go:2628]: It's not obvious to me why this variable is being renamed

File: shared/pingpong/accounts.go [hunk 10]
```diff
@@ -197,9 +197,9 @@ func (pps *WorkerState) ensureAccounts(ac *libgoal.Client) (err error) {
 				srcAcctPresent = true
 			}
 
-			ai, err := ac.AccountInformation(addr, true)
-			if err != nil {
-				return err
+			ai, aiErr := ac.AccountInformation(addr, true)
+			if aiErr != nil {
+				return aiErr
```
Comment (tzaffi) [inline shared/pingpong/accounts.go:202]:
```suggestion
			ai, err := ac.AccountInformation(addr, true)
			if err != nil {
				return
```
Suggestion to stay within style of the rest of this function. Alternately, we might want to root out all the naked returns -but that could be in a follow up as previously discussed.

**Review by tzaffi — COMMENTED** [review]
I agree with the vast majority of the changes. I left a couple of questions/suggestions in places that I disagreed or could spot an alternative.

File: shared/pingpong/accounts.go (hunk 10 above)
Comment (tzaffi) [reply shared/pingpong/accounts.go:202]: (the spacing is all messed up in the suggestion)

File: ledger/catchpointtracker.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (winder) [reply ledger/catchpointtracker.go:1210]: FWIW I like to add scopes like this in long functions. It helps me keep track of where locals are supposed (or prevent them from being used where they shouldn't be, depending on your perspective).

**Review by tzaffi — DISMISSED** [review]
LGTM

**Review by jannotti — APPROVED** [review]
I'll try to look at more later, but this looks mergable to me.

**Review by tzaffi — APPROVED** [review]

**Review by gmalouf — APPROVED** [review]

**Pull request merged**